Add policy delegate for back/forward navigations

beidson opened the pull request:
#### 3555845f0475627457afa3c7b968e783875f1669
<pre>
Add policy delegate for back/forward navigations
<a href="https://rdar.apple.com/3979539">rdar://3979539</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=287182">https://bugs.webkit.org/show_bug.cgi?id=287182</a>

Reviewed by Chris Dumez.

Adds an asynchronous policy delegate for back/forward navigations except for ones
that are required to be synchronous by the web platform.

Do a sync IPC for the policy delegate for those.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::shouldGoToHistoryItem const):
(WebCore::EmptyFrameLoaderClient::supportsAsyncShouldGoToHistoryItem const):
(WebCore::EmptyFrameLoaderClient::shouldGoToHistoryItemAsync const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::FrameLoader):

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::ref const):
(WebCore::HistoryController::deref const):
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::goToItemForNavigationAPI):
(WebCore::HistoryController::goToItemShared):
* Source/WebCore/loader/HistoryController.h:

* Source/WebCore/loader/LocalFrameLoaderClient.h:

* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::shouldGoToBackForwardListItem):
(API::NavigationClient::willGoToBackForwardListItem): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:

* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::setNavigationDelegate):
(WebKit::NavigationState::NavigationClient::shouldGoToBackForwardListItem):
(WebKit::NavigationState::NavigationClient::willGoToBackForwardListItem): Deleted.

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::didReceiveMessage):

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::shouldGoToBackForwardListItem):
(WebKit::WebPageProxy::shouldGoToBackForwardListItemSync):
(WebKit::WebPageProxy::willGoToBackForwardListItem): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItem const):
(WebKit::WebLocalFrameLoaderClient::supportsAsyncShouldGoToHistoryItem const):
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItemAsync const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:

* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::shouldGoToHistoryItem const):
(WebFrameLoaderClient::supportsAsyncShouldGoToHistoryItem const):
(WebFrameLoaderClient::shouldGoToHistoryItemAsync const):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(-[BackForwardDelegateWithShouldGo _webView:willGoToBackForwardListItem:inPageCache:]):
(-[BackForwardDelegateWithShouldGo _webView:shouldGoToBackForwardListItem:inPageCache:completionHandler:]):
(-[BackForwardDelegateWithShouldGo webView:didFinishNavigation:]):
(TEST(WKNavigation, ShouldGoToBackForwardListItem)):

Canonical link: <a href="https://commits.webkit.org/290200@main">https://commits.webkit.org/290200@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92925b3cd290d341037c05aa78e26bc9357f128e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89264 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8788 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44098 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94247 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40025 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91315 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9176 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17007 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68773 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26445 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81013 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49134 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6785 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35394 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39132 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36382 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96079 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16444 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77651 "Failure limit exceed. At least found 1 new test failure: svg/dynamic-updates/SVGFEDiffuseLightingElement-lighting-color-css-prop.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16700 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76800 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76948 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18976 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21362 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19968 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9564 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16458 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21769 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16199 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19650 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17980 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -294,40 +294,51 @@ bool HistoryController::shouldStopLoadingForHistoryItem(HistoryItem& targetItem)
-
-    // Now that all other frames have provisional items, do the actual navigation.
-    recursiveGoToItem(targetItem, currentItem.get(), type, shouldTreatAsContinuingLoad);
+    auto finishGoToItem = [weakFrame = WeakPtr { m_frame.ptr() }, this, frameLoadType, shouldTreatAsContinuingLoad, page, targetItem = Ref { targetItem }] (bool result) {
```
`weakFrame = WeakPtr { m_frame.get() }` is a bit better as it avoids a null check.

cdumez replied:
What guarantees that `this` is still alive when this lambda runs? I assume it can run asynchronously. Odds are that we should have HistoryController forward its refcounting to its owner (FrameLoader) and capture a `protectedThis` or `weakThis` here.

beidson replied:
As you probably found, `HistoryController` is not ref counted or weakptr'ed, and is just owned by the `FrameLoader`

Hence the `RELEASE_ASSERT(&weakFrame->loader().history() == this);` check after verifying the Frame is still valid.

I'll add ref/deref that forwards to the FrameLoader as you suggest.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -294,40 +294,51 @@ bool HistoryController::shouldStopLoadingForHistoryItem(HistoryItem& targetItem)
-    // Now that all other frames have provisional items, do the actual navigation.
-    recursiveGoToItem(targetItem, currentItem.get(), type, shouldTreatAsContinuingLoad);
+    auto finishGoToItem = [weakFrame = WeakPtr { m_frame.ptr() }, this, frameLoadType, shouldTreatAsContinuingLoad, page, targetItem = Ref { targetItem }] (bool result) {
+        // `this` will be valid if the Frame is still valid
```
Please do what I suggest above instead and we can remove this comment. We shouldn't have such fragile patterns in modern code.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -336,49 +347,80 @@ struct HistoryController::FrameToNavigate {
-        if (RefPtr localMainFrame = page->localMainFrame())
-            recursiveGatherFramesToNavigate(*localMainFrame, framesToNavigate, targetItem, fromItem.get());
-    }
+    auto finishGoToItem = [weakFrame = WeakPtr { m_frame.ptr() }, this, frameLoadType, page, triggeringFrame = Ref { triggeringFrame }, targetItem = Ref { targetItem }, tracker = RefPtr { tracker }] (bool result) {
```
Same comments as for the lambda above.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -336,49 +347,80 @@ struct HistoryController::FrameToNavigate {
+        return;
     }
+
+    m_frame->loader().client().shouldGoToHistoryItemAsync(targetItem, [completionHandler = WTFMove(completionHandler)] (bool result) mutable {
```
Can we do this?
```
m_frame->loader().client().shouldGoToHistoryItemAsync(targetItem, WTFMove(completionHandler));
```

beidson replied:
Yes - this weird extra lambda made its way in during an earlier version of the patch, but the need for it has factored out 👍

cdumez commented on Source/WebCore/loader/HistoryController.h:
```diff
@@ -100,11 +100,14 @@ class HistoryController final : public CanMakeCheckedPtr<HistoryController> {
 
     WEBCORE_EXPORT RefPtr<HistoryItem> createItemTree(LocalFrame& targetFrame, bool clipAtTarget, BackForwardItemIdentifier);
 
+    bool isSameDocumentNavigation(HistoryItem&);
```
Where is the implementation for this? May be leftover code.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -1244,15 +1244,38 @@ void WebLocalFrameLoaderClient::updateGlobalHistoryRedirectLinks()
+        return;
+    }
+
+    webPage->sendWithAsyncReply(Messages::WebPageProxy::ShouldGoToBackForwardListItem(item.itemID(), item.isInBackForwardCache()), [completionHandler = WTFMove(completionHandler)] (bool result) mutable {
```
Would this work?
```
webPage->sendWithAsyncReply(Messages::WebPageProxy::ShouldGoToBackForwardListItem(item.itemID(), item.isInBackForwardCache()), WTFMove(completionHandler));
```

beidson replied:
Exact same comment as above. Extra lambda no longer needed.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -294,40 +294,48 @@ bool HistoryController::shouldStopLoadingForHistoryItem(HistoryItem& targetItem)
-
-    // Now that all other frames have provisional items, do the actual navigation.
-    recursiveGoToItem(targetItem, currentItem.get(), type, shouldTreatAsContinuingLoad);
+    auto finishGoToItem = [this, weakThis = WeakPtr { this }, frameLoadType, shouldTreatAsContinuingLoad, page, targetItem = Ref { targetItem }] (bool result) {
```
We should not be capturing both `this` and and `weakThis` per Safer CPP guidelines. It's unfortunate the Safer CPP bot ignores this file for now. We should:
1. Capture `weakThis` only, not `this`
2. Convert the `weakThis` into a `protectedThis` at the beginning of the lambda
3. Use `protectedThis->` whenever necessary (if the logic is too long and that's too many `protectedThis->`, then it indicates we should move that logic to a separate member function.

cdumez replied:
You motivated me to un-skip this file on the Safer CPP bot. I'll try and get it done today.

cdumez replied:
https://github.com/WebKit/WebKit/pull/40319

beidson replied:
As established in comments here and on slack, I went the weak route because the straightforward ref/deref forwarding broke the build with static asserts.

We found the workaround for that.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -336,49 +344,75 @@ struct HistoryController::FrameToNavigate {
-        if (RefPtr localMainFrame = page->localMainFrame())
-            recursiveGatherFramesToNavigate(*localMainFrame, framesToNavigate, targetItem, fromItem.get());
-    }
+    auto finishGoToItem = [this, weakThis = WeakPtr { this }, frameLoadType, page, triggeringFrame = Ref { triggeringFrame }, targetItem = Ref { targetItem }, tracker = RefPtr { tracker }] (bool result) {
```
Same comments apply.

cdumez commented on Source/WebCore/loader/HistoryController.h:
```diff
@@ -45,7 +45,7 @@ enum class ShouldTreatAsContinuingLoad : uint8_t;
 struct StringWithDirection;
 
-class HistoryController final : public CanMakeCheckedPtr<HistoryController> {
+class HistoryController final : public CanMakeCheckedPtr<HistoryController>, public CanMakeWeakPtr<HistoryController> {
```
I see you still did not forward the ref-counting, I think we really should. Normally you'd get a build error if you subclass `CanMakeWeakPtr` without being ref-counted. However, the `CanMakeCheckedPtr` is saving you before in theory, it would be safe (from a security standpoint) to convert the WeakPtr into a CheckedPtr and then only use that CheckedPtr. However, your lambdas just use the WeakPtr directly and never convert to a CheckedPtr or a RefPtr so the code is not safe. I do think that RefPtr would be better than CheckedPtr given that it is trivial to forward the ref-counting here (I think?) and it would avoid crashes if the object gets destroyed in the middle of the lambda running.

beidson replied:
I mentioned this to you on Slack, actually - I cannot forward ref counting without completely reworking a fragile class hierarchy, because HistoryController is currently a UniqueRef, which is incompatible with RefCounted.

beidson replied:
On Slack we established the build break, and discovered the workaround of makeUniqueRefWithoutRefCountedCheck

cdumez replied:
calling `makeUniqueRefWithoutRefCountedCheck` would resolve your build issue when forwarding refcounting and using UniqueRef.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -297,40 +307,50 @@ bool HistoryController::shouldStopLoadingForHistoryItem(HistoryItem& targetItem)
+    auto finishGoToItem = [weakThis = WeakPtr { this }, frameLoadType, shouldTreatAsContinuingLoad, page, targetItem = Ref { targetItem }] (bool result) {
+        if (!weakThis)
+            return;
+        Ref protectedThis = Ref { *weakThis };
```
I know this doesn't matter much in single threaded code but our usual pattern is:
```
RefPtr protectedThis = weakThis.get();
if (!protectedThis)
    return;
```

Which is safe even in multi-threaded code.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -339,49 +359,77 @@ struct HistoryController::FrameToNavigate {
+    auto finishGoToItem = [weakThis = WeakPtr { this }, frameLoadType, page, triggeringFrame = Ref { triggeringFrame }, targetItem = Ref { targetItem }, tracker = RefPtr { tracker }] (bool result) {
+        if (!weakThis)
+            return;
+        Ref protectedThis = Ref { *weakThis };
```
ditto.

cdumez commented on Source/WebCore/loader/LocalFrameLoaderClient.h:
```diff
@@ -217,7 +217,9 @@ class WEBCORE_EXPORT LocalFrameLoaderClient : public FrameLoaderClient {
     virtual void updateGlobalHistoryRedirectLinks() = 0;
 
-    virtual bool shouldGoToHistoryItem(HistoryItem&) const = 0;
+    virtual bool shouldGoToHistoryItem(HistoryItem&, bool sameDocumentNavigation) const = 0;
```
Ideally, the second parameter should really be an enum class:
```
enum class IsSameDocumentNavigation : bool { No, Yes };
```

as we normally frown upon such boolean parameters, which can make call sites less readable.

cdumez reviewed (approved):
LGTM with nits.

cdumez commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -339,50 +359,80 @@ struct HistoryController::FrameToNavigate {
+        return;
     }
+
+    m_frame->protectedLoader()->protectedClient()->shouldGoToHistoryItemAsync(targetItem, WTFMove(completionHandler));
```
Should be using `frame` instead of `m_frame` here per the Safer CPP bot.

beidson committed:
Add policy delegate for back/forward navigations
rdar://3979539
https://bugs.webkit.org/show_bug.cgi?id=287182

Reviewed by Chris Dumez.

Adds an asynchronous policy delegate for back/forward navigations except for ones
that are required to be synchronous by the web platform.

Do a sync IPC for the policy delegate for those.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::shouldGoToHistoryItem const):
(WebCore::EmptyFrameLoaderClient::supportsAsyncShouldGoToHistoryItem const):
(WebCore::EmptyFrameLoaderClient::shouldGoToHistoryItemAsync const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::FrameLoader):

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::ref const):
(WebCore::HistoryController::deref const):
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::goToItemForNavigationAPI):
(WebCore::HistoryController::goToItemShared):
* Source/WebCore/loader/HistoryController.h:

* Source/WebCore/loader/LocalFrameLoaderClient.h:

* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::shouldGoToBackForwardListItem):
(API::NavigationClient::willGoToBackForwardListItem): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationDelegatePrivate.h:

* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::setNavigationDelegate):
(WebKit::NavigationState::NavigationClient::shouldGoToBackForwardListItem):
(WebKit::NavigationState::NavigationClient::willGoToBackForwardListItem): Deleted.

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::didReceiveMessage):

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::shouldGoToBackForwardListItem):
(WebKit::WebPageProxy::shouldGoToBackForwardListItemSync):
(WebKit::WebPageProxy::willGoToBackForwardListItem): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItem const):
(WebKit::WebLocalFrameLoaderClient::supportsAsyncShouldGoToHistoryItem const):
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItemAsync const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:

* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::shouldGoToHistoryItem const):
(WebFrameLoaderClient::supportsAsyncShouldGoToHistoryItem const):
(WebFrameLoaderClient::shouldGoToHistoryItemAsync const):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(-[BackForwardDelegateWithShouldGo _webView:willGoToBackForwardListItem:inPageCache:]):
(-[BackForwardDelegateWithShouldGo _webView:shouldGoToBackForwardListItem:inPageCache:completionHandler:]):
(-[BackForwardDelegateWithShouldGo webView:didFinishNavigation:]):
(TEST(WKNavigation, ShouldGoToBackForwardListItem)):

Canonical link: https://commits.webkit.org/290200@main
